### PR TITLE
Prevent multiple messages for one violation

### PR DESCRIPTION
**Pull request by yashincontrol**
Fixes https://github.com/iTwin/eslint-plugin/issues/22. The problem is that the no-internal rule was generating duplicate error messages. This was because the rule was not correctly handling the traversal of declarations and their parents, leading to redundant checks and violation messages. Fixed the function calls and added a `visitedDeclarations` set to avoid duplicate processing of declarations.

### CODE REVIEW

**Review by ben-polinsky — DISMISSED** [review]

File: tests/no-internal.js [hunk 1]
```diff
@@ -61,14 +61,7 @@ ruleTester.run(
         `,
         errors: [
           { message: 'function "internal" is internal.' },
-          // FIXME: there is separate work for removing duplicate violations
           { message: 'class "Internal" is internal.' },
```
Comment (MichaelBelousov) [inline tests/no-internal.js:64]: is it just me or does this remove too many instances?  I see 3 remaining but shouldn't there be 4? (`internal()`, `new Internal()`, `new Internal().publicMethod()`, `new Public().internalMethod()`)?

Comment (yashincontrol) [reply tests/no-internal.js:64]: Yes, I misinterpreted the test file. Will try to understand the no-internal rule better and debug

**Review by ben-polinsky — COMMENTED** [review]
Looks better. 
In general, the tests are a bit confusing due to the @bentley/@itwin scope rule. Could you add a comment next to the import(s) clarifying why the code is valid or invalid? Something like "not a bentley/itwin scope" or "is a local import"

**Review by ben-polinsky — APPROVED** [review]

**Comment by yashincontrol** [comment]
> Looks better. In general, the tests are a bit confusing due to the @bentley/@iTwin scope option/requirement. Could you add a comment next to the import(s) clarifying why the code is valid or invalid? Something like "not a bentley/itwin scope" or "is a local import"

Yes, will add that

File: tests/no-internal.js (hunk 1 above)
Comment (MichaelBelousov) [reply tests/no-internal.js:66]: looks correct to me now, thanks!

File: dist/rules/no-internal.js [hunk 2]
```diff
@@ -171,11 +176,11 @@ module.exports = {
 
         const resolvedClass = typeChecker.getTypeAtLocation(tsCall);
         if (resolvedClass && resolvedClass.symbol)
-          checkJsDoc(resolvedClass.symbol.valueDeclaration, node);
-
+          checkWithParent(resolvedClass.symbol.valueDeclaration, node);
+
```
Comment (MichaelBelousov) [inline dist/rules/no-internal.js:180]: excess spaces on this line

File: tests/no-internal.js [hunk 3]
```diff
@@ -30,13 +30,14 @@ const ruleTester = new ESLintTester({
...
 
 ruleTester.run(
-  "no-internal-barrel-imports",
+  "no-internal",
   NoInternalESLintRule,
   supportSkippedAndOnlyInTests({
     valid: [
-      { code: `import { internal, public, Internal, Public } from "test-pkg-1";` },
-      { code: `import * as Local from "./local-internal"; Local.internal(); new Local.Internal();` },
+      { code: `import { internal, public, Internal, Public } from "test-pkg-1";` }, //not a bentley/itwin scope
```
Comment (MichaelBelousov) [inline tests/no-internal.js:37]: nit:
```suggestion
      { code: `import { internal, public, Internal, Public } from "test-pkg-1";` }, // not a bentley/itwin scope
```

File: tests/no-internal.js [hunk 4]
```diff
@@ -30,13 +30,14 @@ const ruleTester = new ESLintTester({
...
+      { code: `import * as Local from "./local-internal"; Local.internal(); new Local.Internal();` }, //local import
```
Comment (MichaelBelousov) [inline tests/no-internal.js:38]: nit:
```suggestion
      { code: `import * as Local from "./local-internal"; Local.internal(); new Local.Internal();` }, // local import
```

File: tests/no-internal.js [hunk 5]
```diff
@@ -61,20 +63,14 @@ ruleTester.run(
...
-          { message: 'class "Internal" is internal.' },
-          { message: 'method "Public.internalMethod" is internal.' },
-          { message: 'method "Public.internalMethod" is internal.' },
           { message: 'method "Public.internalMethod" is internal.' }
         ]
       },
       {
         code: dedent`
-          import { internal, public, Internal, Public } from "test-pkg-1";
+          import { internal, public, Internal, Public } from "test-pkg-1";
```
Comment (MichaelBelousov) [inline tests/no-internal.js:73]: excess space at end
```suggestion
          import { internal, public, Internal, Public } from "test-pkg-1";
```

File: tests/no-internal.js [hunk 6]
```diff
@@ -50,6 +51,7 @@ ruleTester.run(
     ],
     invalid: [
       {
+        //itwin scope
```
Comment (MichaelBelousov) [inline tests/no-internal.js:54]:
```suggestion
        // itwin scope
```

File: tests/no-internal.js [hunk 7]
```diff
@@ -30,13 +30,14 @@ const ruleTester = new ESLintTester({
...
+  "no-internal",
   NoInternalESLintRule,
   supportSkippedAndOnlyInTests({
     valid: [
-      { code: `import { internal, public, Internal, Public } from "test-pkg-1";` },
-      { code: `import * as Local from "./local-internal"; Local.internal(); new Local.Internal();` },
+      { code: `import { internal, public, Internal, Public } from "test-pkg-1";` }, //not a bentley/itwin scope
+      { code: `import * as Local from "./local-internal"; Local.internal(); new Local.Internal();` }, //local import
       {
+        //not a bentley/itwin scope
```
Comment (MichaelBelousov) [inline tests/no-internal.js:40]:
```suggestion
        // not a bentley/itwin scope
```

**Review by MichaelBelousov — APPROVED** [review]
lgtm, minor comments mostly formatting nits which can be resolved at your leisure

**Comment by MichaelBelousov** [comment]
> Looks better. In general, the tests are a bit confusing due to the @bentley/@iTwin scope option/requirement. Could you add a comment next to the import(s) clarifying why the code is valid or invalid? Something like "not a bentley/itwin scope" or "is a local import"

Great idea, does make them much more readable.

File: dist/rules/no-internal.js (hunk 2 above)
Comment (yashincontrol) [reply dist/rules/no-internal.js]: fixed

File: tests/no-internal.js (hunk 3 above)
Comment (yashincontrol) [reply tests/no-internal.js]: Fixed

File: tests/no-internal.js (hunk 4 above)
Comment (yashincontrol) [reply tests/no-internal.js]: fixed

File: tests/no-internal.js (hunk 5 above)
Comment (yashincontrol) [reply tests/no-internal.js]: fixed

File: tests/no-internal.js (hunk 6 above)
Comment (yashincontrol) [reply tests/no-internal.js]: fixed

File: tests/no-internal.js (hunk 7 above)
Comment (yashincontrol) [reply tests/no-internal.js]: fixed

**Pull request merged**